compiler/rustc_ty_utils/src/abi.rs

```diff
@@ -822,6 +835,44 @@ fn fn_abi_adjust_for_abi<'tcx>(
     Ok(())
 }
 
+fn can_annotate_small_immediate_argument_with_noundef<'tcx>(
```

Comment:
I'm not super confident in the correctness of this eligibility check here. I've added positive and negative test cases for edge cases like unions or array of unions, but I still feel like I'm missing cases.